# Unit 2 - In Class: IF statement and Operators

### Overview of Lesson
* Sequence
* Selection
* IF statement - `IF` or `ELSE`
* Comparison vs Assignment
* Indent in python
* `>` vs `>=`

* What are Operators?  - https://www.youtube.com/watch?v=Y6CwThhquQs&t=119s&ab_channel=GoogleStudents

### if/else statement (String)

```python
if lastName=="Bob":
  print("Welcome Bob!")
else:
  print("you are not Bob!")
```

### if/elif/else (int)
```python
if age>14:
  print("Age is above 14")
elif age>10 and age<=14:
  print("Age is between 10 and 14")
else:
  print("Under 14")
```


### Logical Operators

* `==` Equal to, `5==4` -> `False`
* `!=` Not equal to, `5!=4` -> `True`
* `<`  Less than, 2<10 -> `True`
* `<=` Less than or equal to, 2<=10 -> `True`
* `>` Greater than, 9>9 -> `False`
* `>=` Greater than or equal to, 9>=9 -> `True`
* `and` AND, 3>2 and 5==4 -> `False`
* `or` OR, 3>2 or 5==4 -> `True`

examples of logical operators

```
x != y               # x is not equal to y
x > y                # x is greater than y
x < y                # x is less than y
x >= y               # x is greater than or equal to y
x <= y               # x is less than or equal to y
```

## Helpful resources
* https://docs.python.org/3/reference/expressions.html#expression-lists

---
## Agenday

This notebook covers the following concepts:

* Variables
* Data types
* Operators


---
## Simple Arithmetic operators examples & Recap from Lesson 1 (Variables & Data Types)

about arithmetic operators:
* Arithmetic operators are used to perform mathematical operations on data.
* There are many different arithmetic operators in Python, but some of the most common ones include addition, subtraction, multiplication, division, and exponentiation.
* The order of operations in Python is important. This means that operations are performed in a specific order, starting with the operations inside parentheses, then exponentiation, then multiplication and division, and then addition and subtraction.
* When concatenating strings, it is important to use the correct type of data. For example, you cannot concatenate a string and a number.

In [13]:
print(34 + 12)

print("34" + "12") # concatenation - make sure to use the right type

46
3412


The first line of code, `print(34 + 12)`, uses the addition operator to add the two numbers 34 and 12. The result of the addition is 46.

The second line of code, `print("34" + "12")`, uses the concatenation operator to concatenate the two strings "34" and "12". The result of the concatenation is the string "3412".

The difference between the two lines of code is that the first line adds two numbers, while the second line concatenates two strings.

----
lets try another example:

In [15]:
age = input("How old are you? ")
age+1
print(age)

TypeError: can only concatenate str (not "int") to str

we use `age = input("How old are you? ")`, as the `input` function to get the user's age as input. The `input` function returns a string, so the value of the variable age is a string.

This will throw an error because the value of the variable `age` is a `string`, not an `integer`. The `+` operator cannot be used to add a string and an integer.

To fix the error, we need to convert the value of the variable age to an integer. We can do this using the `int()` function. The following code will work:

In [16]:
age = int(input("How old are you? ")) # 15
age+1
print(age)

15


notice `age+1`, adds 1 to the value of the variable age. Since the value of the variable age is now an integer, the addition operation succeeds and the value of the variable age becomes 15. __but not 16__ because we didnt save the value of `age+1` to the variable `age`. The following code saves the value of `age+1` to the variable `age`:

In [17]:
age = age+1
print(age)

16


---

Now lets try another example:
* allow the user to `input` two numbers, `n1` and `n2`
* output the following
  * n1 add n2 (+)
  * n1 substract n2 (-)
  * n1 multiply n2 (*)

In [19]:
n1 = int(input("What is the first number? "))
n2 = int(input("What is the second number? "))
print("The sum of the two numbers is: "+ str(n1+n2))
print("substract the first number from the second: "+ str(n1-n2))
print("multiplied together: "+ str(n1*n2))

The sum of the two numbers is: 9
substract the first number from the second: 1
multiplied together: 20


---
## IF statement

An `if` statement in Python is a conditional statement that allows you to execute code based on a condition. The syntax for an if statement is as follows:

```python
if condition:
  statement(s)
```

The if statement will only execute the statements inside the if block if the condition is true. If the condition is false, the if block will be skipped. for example:

```python
#if/else statement (String)
if lastName=="Bob":
  print("Welcome Bob!")
elif lastName=="Robert":
  print("Welcome Robert!, do you go by Bob?")
else:
  print("you are not Bob!")
```

### Key Points about IF statements and conditions
* `if` statements are a very important part of programming. They allow you to control the flow of your code based on different conditions.
* There are other types of **conditional **statements in Python, such as the `elif` statement and the else statement.
* The `elif` statement allows you to check for multiple conditions.
* The `else` statement allows you to execute code if none of the other conditions are met.

In [ ]:
print("Lets make a PB&J sandwich")
print("Get two slices of bread")
print("Get peanut butter")
print("Get jelly")
print("Spread peanut butter on one slice of bread")
print("Spread jelly on the other slice of bread")
print("Put the two slices of bread together")
print("Enjoy your sandwich")

What if we dont want `jelly`?

In [20]:
print("Lets make a PB&J sandwich")
print("Get two slices of bread")
print("Get peanut butter")
print("Spread peanut butter on one slice of bread")
jelly = input("Do you have jelly? ")
if jelly == "yes":
    print("Get jelly")
    print("Spread jelly on the other slice of bread")
else:
    print("add more peanut butter 😄")
print("Put the two slices of bread together")
print("Enjoy your sandwich")

Lets make a PB&J sandwich
Get two slices of bread
Get peanut butter
Spread peanut butter on one slice of bread
add more peanut butter 😄
Put the two slices of bread together
Enjoy your sandwich


what `if` we want more sandwiches? but we only have 7 slices of bread?

In [24]:
#what `if` we want more sandwiches? but we only have 7 slices of bread?
bread_num = 7
sandwich_num = int(input("How many sandwhiches do you want? "))

if sandwich_num <= bread_num//2:
    print("we have enough bread to make "+ str(sandwich_num) + " sandwiches")
    print("Lets make a PB&J sandwich")
else:
    print("Sorry, we don't have enough bread")


we have enough bread to make 3 sandwiches
Lets make a PB&J sandwich


explaination to the example:

The code will only print the text "Lets make a PB&J sandwich" if the user enters a number of sandwiches that is less than or equal to half of the number of slices of bread we have. If the user enters a number of sandwiches that is greater than half of the number of slices of bread we have, the code will print the text "Sorry, we don't have enough bread".

`if sandwich_num <= bread_num//2`, is an if statement. The condition of the `if` statement is `sandwich_num <= bread_num//2`. This means that the if statement will only be executed if the number of sandwiches the user wants to make is less than or equal to half of the number of slices of bread we have.

---
## Activity 1
- A student takes three test, each test is out of 20. Allow the student to enter their test score.
- If the student scores a total of 36, or more, they have passed, otherwise they have failed. 
- Output if the student has passes or failed. 

> Go to the `lesson2_activity` notebook

In [25]:
## part 1: request students to enter their 3 test scores
test1 = int(input("Please input test score 1: "))
test2 = int(input("Please input test score 2: "))
test3 = int(input("Please input test score 3: "))

In [26]:
## part 2: find the total scores
totalScore = test1 + test2 + test3

In [27]:
## part 3: if statement (if total score is more than 36 pass if not failed) print out passes and failures
if totalScore >= 36:
    print("You passed the test!")
else:
    print("You failed the test!")

You failed the test!


---
## Operators

* Boolean - True / False
* Making use of OR and AND logical operators
* Nesting vs AND
* Using != as not equals


Key Points
* The or operator is a logical operator that evaluates to True if one or both of the operands are True.
* The and operator is a logical operator that evaluates to True if both of the operands are True.
* The !== operator is the not equal to operator. It evaluates to True if the two operands are not equal.

In [28]:
T = 15
S = 12
F = True

if 15 > 12 or True and 15 != 14:
  print("y")

y


above `if 15 > 12 or True and 15 != 14:`, is an if statement. The condition of the `if` statement is `15 > 12 or True and 15 != 14`. This means that the if statement will only be executed if one of the following conditions is true:

* 15 is greater than 12
* The value of the variable `F` is True
* The value of the variable `T` is not equal to 14
* If one of the conditions is true, the code inside the if block will be executed. The code inside the if block prints the text "y".

If none of the conditions is true, the code inside the if block will not be executed.

The code will print the text "y" if one of the following conditions is true:

* 15 is greater than 12 (which is true)
* The value of the variable F is True (which is also true)
* The value of the variable T is not equal to 14 (which is also true)

---
class example 
* store the variables:
  * Username = 'Bob'
  * Password = 'Rocks'
* ask the end user to inpu their username. If the username matches "Bob" then allow the password tto be entered. 
* Output the message "Access Granted" or "Wrong user name or password"

In [ ]:
# store the variables
Username = 'Bob'
Password = 'Rocks'

# ask the end user to input their username
username = input("Enter your username: ")

# if the username matches "Bob" then allow the password to be entered
if username == 'Bob':
    password = input("Enter your password: ")

    # if the password matches "Rocks" OR 'roxs' then output "Access Granted"
    if password == 'Rocks' or password == 'roxs':
        print("Access Granted")
    else:
        print("Wrong password")
else:
    print("Wrong username")

---
## Activity 2
> Go to the `lesson2_activity` notebook

**Part 1**
extend the previous activity and output the following grade:
* 48 - Perfect 
* 42 - Excellent
* 36 - Pass
* 33 - Good
* 30 - Needs Improvement
* 27 - Needs to Work Harder

**Part 2** 
* using the initial test scores from `test1`, `test2` and `test3` uses operators such as `and`, `or`, `and not` to check multiple conditions, of the following:
* if any of the test1, test2 or test3 are below 10, output the message "You have failed the exam"
* if test1 is greater than 10 and test2 is greater than 10 and test3 is greater than 10, output the message "You have passed the exam"
* if test2 or test3 are greater than 10, output the message "those were tough questions, but you have passed the exam"
